### PR TITLE
Update tx with missing fields.

### PR DESCRIPTION
**Pull request by zmrocze**
`Tx` lacks metadata/withdrawals/certificates fields #471. 
Together with adding these I refactored script and datum witnesses out of inputs/etc into seperate maps [see #459 ]. 

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [ ] Reviewer requested

This pr introduces these changes:

### Redefines `Tx` type
 - `Tx` gets withdrawals, certificates(1), metadata(1) fields
 - script and datum witnesses get put into maps and refered to by hashes ([related comment](https://github.com/input-output-hk/plutus-apps/issues/471#issuecomment-1133941543) and [issue](#459) )
 - redeemers get put together with the MintingPolicy/Certificate/Input (^ same links)
 - defines `TxInput` and `TxInputType` simmilar to `TxIn` and `TxInType`.
 - defines `Withdrawal` type
 - defines `Certificate` type
 - splits Ledger.Tx.Internal module

### Updates pab accordingly
 - `TxIn` remains used by some functions only in auxiliary way
 - most changes are automatic: imports and minor updates that don't change logic only update the `Tx`
 - `toCardanoTxBodyContent` from Ledger.Tx.CardanoAPI gets updated with withdrawal translation (based on [plutus-simple-model](https://github.com/mlabs-haskell/plutus-simple-model/blob/main/src/Plutus/Test/Model/Fork/CardanoAPI.hs)).

### Updates purescript accordingly
 - code generation generates added types: `Certificate`, `Withdrawal`, `TxInput`, `TxIn`
 - minor tweak in purescript source code

### Certificate and Metadata is wrong for now (1)
 - `DCert` in `Certificate` has some data like `StakePoolParameters` in registration certificate digested. Best would be to reproduce [cardano-api/`Certificate`](https://github.com/input-output-hk/cardano-node/blob/6a465b09b5dfd5fe281061b94f69faee1227800a/cardano-api/src/Cardano/Api/Certificate.hs#L80) but using plutus-friendly types.
 - Metadata is more like `Map Word64 JSONLikeValue`, definitely not bytestring. Best would be [cardano-api/`TxMetadata`](https://github.com/input-output-hk/cardano-node/blob/6a465b09b5dfd5fe281061b94f69faee1227800a/cardano-api/src/Cardano/Api/TxMetadata.hs#L83), but we can't use for example Word64 onchain, correct? We could reproduce cardano-api type using plutus-friendly types. Or leave metadata as bytestring, to be decoded by users (I don't like this).

### What about plutus V2? i.e Should scripts in txScripts be annotated with versions?
 - EDIT: I see that script version was added to TxInType

### CODE REVIEW

**Comment by koslambrou** [comment]
Can you rebase from `next-node` branch? There are a lot of commits in here that are unrelated to the work done.

**Comment by zmrocze** [comment]
^ Yes, that was my mistake. I was baited by `Tx` moved inside `plutus-ledger` in `next-node`.

I rebased it on main. Still need to fix CI etc.

**Comment by koslambrou** [comment]
@zmrocze Just noticed a huge mistake on my part... It was actually okay and desirable for you to work on `next-node`. I had forgotten that the `Tx` is available in `main` branch, and was migrate only in `next-node`. So this PR is the right one. Really sorry about getting mixed up...

**Comment by zmrocze** [comment]
oh damn..

@koslambrou  I'll go about resolving conflicts here, hopefully not many. I'm close to resolving them as well in the main targeting branch, if thats relevant.

I would still need advice on:
 - what to use for `DCert` and `Metadata` types? Should we define in Ledger analogues of Cardano.Api types, just using Plutus friendly type analogues? Or use Cardano.Api types itself (not plutus friendly).
 
 Maybe I could correct `Metadata` already in this pr, as I've seen someone having to use a fork with metadata added. But leave `DCert` for next pr already, not to leave this hanging for longer?
 
 I've edited slightly the pr description above but it mainly remains up to date.

**Comment by koslambrou** [comment]
No, don't waste time on the `main`  branch. Let's stay on `next-node` unless you have a compelling reason why you would want it also on `main`.

For the types, let's use the ones from `Cardano.Api`. We're moving towards using that more and more.

And yeah, I would agree to add Metadata only in this PR, and add DCert in a separate one.

**Comment by zmrocze** [comment]
@koslambrou  I think its ready for review.

I had to reset back to commit without TxMetadata used in Tx as I need help fixing:
```
[1/1 ModuleNotFound] generated/Ledger/Tx/Internal.purs:6:1

  6  import Cardano.Api.TxMetadata (TxMetadata)
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  Module Cardano.Api.TxMetadata was not found.
  Make sure the source file exists, and that it has been provided as an input to the compiler.
```
that is `plutus-playground-cllient/default.nix:generate-purescript` needs Cardano.Api as input and I am not fluent with nix enough to fix it.

What is left after that pr:
 - Add constraints in `plutus-ledger-constraints` concerning `Withdrawals`, `Certificate`, `Metadata` (functional constraint)
 - Swap `DCert` in the `txCertificates` field of `Tx` for i.e. cardano-api/`Certificate`.
 - Use cardano-api `TxMetadata` in `txMetadata` field of `Tx` 
 - `TODO: Implement ToSchema TxMetadataValue`
 - `TODO: Parse certifying redeemer.`, `TODO: Parse rewarding redeemer.`, `TODO: ToJSON rewarding/certifying Redeemer.`

Though I think its right to review it already as by its nature it modifies code all over the place and at merges there is a flood of nontrivial merge conflicts.

Initial pr description applies. I tried keeping stuff in `Ledger/Tx/Internal.hs` this time.

**Comment by koslambrou** [comment]
Your PR includes commits that shouldn't be there. Did you do `git pull --rebase origin next-node`?

**Comment by zmrocze** [comment]
I started from next-node when it was WIP pr. Then merged next-node to stay updated. Can `git squash` in the end

File: plutus-ledger-constraints/src/Ledger/Constraints/OffChain.hs [hunk 1]
```diff
@@ -16,6 +16,7 @@
 {-# LANGUAGE TypeOperators             #-}
 {-# LANGUAGE UndecidableInstances      #-}
 {-# LANGUAGE ViewPatterns              #-}
+{-# OPTIONS_GHC -Wno-unused-imports #-}
```
Comment (sjoerdvisscher) [inline plutus-ledger-constraints/src/Ledger/Constraints/OffChain.hs:19]: Why was this added? Can't we just remove the unused imports?

File: plutus-ledger/src/Ledger/Tx.hs [hunk 2]
```diff
@@ -60,41 +59,53 @@ module Ledger.Tx
...
 import Ledger.Validation qualified
 import Plutus.Script.Utils.Scripts (datumHash)
 import Plutus.V1.Ledger.Api qualified as V1
 import Prettyprinter (Pretty (pretty), braces, colon, hang, nest, viaShow, vsep, (<+>))
 
 -- for re-export
-import Plutus.V1.Ledger.Tx hiding (TxIn (..), TxInType (..), inRef, inScripts, inType, pubKeyTxIn, pubKeyTxIns,
-                            scriptTxIn, scriptTxIns)
-import Plutus.V1.Ledger.Tx qualified as V1.Tx hiding (TxIn (..), TxInType (..))
+-- for re-export
```
Comment (sjoerdvisscher) [inline plutus-ledger/src/Ledger/Tx.hs:91]: Something went wrong here. The line is duplicated and we're not exporting `Set`.

File: plutus-ledger/src/Ledger/Tx.hs [hunk 3]
```diff
@@ -319,21 +333,29 @@ getCardanoTxRedeemers = onCardanoTx txRedeemers (const Map.empty) -- TODO: To im
...
+                , hang 2 (vsep ("reference inputs:" : fmap pretty _txReferenceInputs))
+                , hang 2 (vsep ("collateral inputs:" : fmap pretty _txCollateral))
+                , hang 2 (vsep ("outputs:" : fmap pretty _txOutputs))
+                , "mint:" <+> pretty _txMint
+                , "fee:" <+> pretty _txFee
+                , hang 2 (vsep ("mps:": fmap pretty (Map.assocs _txMintingScripts)))
+                , hang 2 (vsep ("signatures:": fmap (pretty . fst) (Map.toList _txSignatures)))
+                , "validity range:" <+> viaShow _txValidRange
+                , hang 2 (vsep ("data:": fmap (pretty . snd) (Map.toList _txData)))
+                , hang 2 (vsep ("attached scripts:": fmap pretty (Map.keys _txScripts)))
```
Comment (sjoerdvisscher) [inline plutus-ledger/src/Ledger/Tx.hs:355]: I think we should only show these 4 lines when they are not empty, this adds quite a bit of noise to the logs.

**Review by sjoerdvisscher — CHANGES_REQUESTED** [review]

File: plutus-ledger-constraints/src/Ledger/Constraints/OnChain/V1.hs [hunk 4]
```diff
@@ -67,7 +67,7 @@ checkOwnOutputConstraint ctx@ScriptContext{scriptContextTxInfo} ScriptOutputCons
     let hsh = V.findDatumHash (Ledger.Datum $ toBuiltinData ocDatum) scriptContextTxInfo
         checkOutput TxOut{txOutValue, txOutDatumHash=Just svh} =
                Ada.fromValue txOutValue >= Ada.fromValue ocValue
-            && Ada.fromValue txOutValue <= Ada.fromValue ocValue + Ledger.maxMinAdaTxOut
+            && Ada.fromValue txOutValue <= Ada.fromValue ocValue + Ledger.minAdaTxOut
```
Comment (sjoerdvisscher) [inline plutus-ledger-constraints/src/Ledger/Constraints/OnChain/V1.hs:70]: Why was this changed?

Comment (zmrocze) [reply plutus-ledger-constraints/src/Ledger/Constraints/OnChain/V1.hs:70]: mistake, corrected

File: plutus-ledger/src/Ledger/Tx.hs (hunk 3 above)
Comment (zmrocze) [reply plutus-ledger/src/Ledger/Tx.hs:355]: metadata line as well, right?

File: plutus-ledger-constraints/src/Ledger/Constraints/OffChain.hs (hunk 1 above)
Comment (zmrocze) [reply plutus-ledger-constraints/src/Ledger/Constraints/OffChain.hs]: Fixed by 'Fix imports.'

File: plutus-ledger/src/Ledger/Tx.hs (hunk 2 above)
Comment (zmrocze) [reply plutus-ledger/src/Ledger/Tx.hs]: Fixed by "Fix imports."

File: plutus-ledger-constraints/src/Ledger/Constraints/OnChain/V1.hs [hunk 5]
```diff
@@ -118,7 +118,7 @@ checkTxConstraint ctx@ScriptContext{scriptContextTxInfo} = \case
             addr = Address.scriptHashAddress vlh
             checkOutput TxOut{txOutAddress, txOutValue, txOutDatumHash=Just svh} =
                    Ada.fromValue txOutValue >= Ada.fromValue vl
-                && Ada.fromValue txOutValue <= Ada.fromValue vl + Ledger.maxMinAdaTxOut
```
Comment (sjoerdvisscher) [inline plutus-ledger-constraints/src/Ledger/Constraints/OnChain/V1.hs:121]: One more here.

File: plutus-contract/src/Plutus/Contract/Wallet.hs [hunk 6]
```diff
@@ -118,26 +119,29 @@ getUnspentOutput = do
     let constraints = mustPayToPubKey ownPkh (Ada.lovelaceValueOf 1)
     utx <- either (throwing _ConstraintResolutionContractError) pure (mkTx @Void mempty constraints)
     tx <- Contract.adjustUnbalancedTx utx >>= Contract.balanceTx
-    case fmap Semigroup.getMin $ foldMap (Just . Semigroup.Min) $ getCardanoTxInputs tx of
-        Just inp -> pure $ txInRef inp
-        Nothing  -> throwing _OtherContractError "Balanced transaction has no inputs"
+    case getCardanoTxInputs tx of
+        inp : _ -> pure $ txInRef inp
+        []      -> throwing _OtherContractError "Balanced transaction has no inputs"
```
Comment (koslambrou) [inline plutus-contract/src/Plutus/Contract/Wallet.hs:124]: What motivated this change?

File: plutus-ledger/src/Ledger/Generators.hs [hunk 7]
```diff
@@ -195,20 +197,23 @@ genValidTransaction' g (Mockchain _ ops _) = do
...
 
 genValidTransactionSpending :: MonadGen m
-    => [TxIn]
+    => [TxInputWitnessed]
     -> Value
     -> m CardanoTx
 genValidTransactionSpending = genValidTransactionSpending' generatorModel
 
+-- | A transaction input, consisting of a transaction output reference and an input type with data witnesses.
+data TxInputWitnessed = TxInputWitnessed !TxOutRef !Ledger.TxInType
```
Comment (koslambrou) [inline plutus-ledger/src/Ledger/Generators.hs:212]: What motivated this change?

File: plutus-ledger/src/Ledger/Tx.hs (hunk 3 above)
Comment (zmrocze) [reply plutus-ledger/src/Ledger/Tx.hs]: Fixed by "Pretty print if non-empty."

File: plutus-ledger-constraints/src/Ledger/Constraints/OnChain/V1.hs (hunk 5 above)
Comment (zmrocze) [reply plutus-ledger-constraints/src/Ledger/Constraints/OnChain/V1.hs]: Fixed by "Correct to use maxMinAdaTxOut. 2".

File: plutus-ledger/src/Ledger/Tx.hs [hunk 8]
```diff
@@ -81,20 +83,28 @@ import Ledger.Address (Address, PaymentPubKey, StakePubKey, pubKeyAddress)
...
 
-type PrivateKey = Crypto.XPrv
+-- | A 'TxOut' along with the 'Tx' it comes from, which may have additional information e.g.
+-- the full data script that goes with the 'TxOut'.
+data TxOutTx = TxOutTx { txOutTxTx :: Tx, txOutTxOut :: TxOut }
+    deriving stock (Show, Eq, Generic)
+    deriving anyclass (Serialise, ToJSON, FromJSON)
+
+txOutTxDatum :: TxOutTx -> Maybe Datum
+txOutTxDatum (TxOutTx tx out) = txOutDatum out >>= (`Map.lookup` txData tx)
```
Comment (koslambrou) [inline plutus-ledger/src/Ledger/Tx.hs:107]: Why was this moved here?

File: plutus-ledger/src/Ledger/Tx/CardanoAPI.hs [hunk 9]
```diff
@@ -9,7 +9,8 @@
 {-# LANGUAGE RecordWildCards    #-}
 {-# LANGUAGE ViewPatterns       #-}
 
-{-# OPTIONS_GHC -Wno-orphans        #-}
+{-# OPTIONS_GHC -Wno-orphans    #-}
+{-# LANGUAGE LambdaCase         #-}
```
Comment (koslambrou) [inline plutus-ledger/src/Ledger/Tx/CardanoAPI.hs:13]: Move this up

File: plutus-ledger/src/Ledger/Tx/CardanoAPI.hs [hunk 10]
```diff
@@ -106,13 +107,17 @@ import Data.Tuple (swap)
...
 import Ledger.Ada qualified as Ada
+import Ledger.Address qualified as L
 import Ledger.Address qualified as P
 import Ledger.Params qualified as P
+import Ledger.Scripts qualified as L
 import Ledger.Scripts qualified as P
+import Ledger.Slot qualified as L
 import Ledger.Slot qualified as P
 import Ledger.Tx.CardanoAPITemp (makeTransactionBody')
+import Ledger.Tx.Internal qualified as L
```
Comment (koslambrou) [inline plutus-ledger/src/Ledger/Tx/CardanoAPI.hs:118]: Duplicated

Comment (koslambrou) [reply plutus-ledger/src/Ledger/Tx/CardanoAPI.hs:118]: Same for the ones above

File: plutus-ledger/src/Ledger/Tx/Internal.hs [hunk 11]
```diff
@@ -70,15 +90,72 @@ instance Pretty TxIn where
...
+data TxInputType =
+      TxConsumeScriptAddress !LedgerPlutusVersion !Redeemer !ValidatorHash !DatumHash -- ^ A transaction input that consumes a script address with the given validator, redeemer, and datum.
+    | TxConsumePublicKeyAddress -- ^ A transaction input that consumes a public key address.
+    | TxConsumeSimpleScriptAddress -- ^ Consume a simple script
+    deriving stock (Show, Eq, Ord, Generic)
+    deriving anyclass (ToJSON, FromJSON, Serialise, NFData)
+
+-- | A transaction input, consisting of a transaction output reference and an input type.
+-- Differs with TxIn by: TxIn *maybe* contains *full* data witnesses, TxInput always contains redeemer witness, but datum/validator hashes.
+data TxInput = TxInput {
```
Comment (koslambrou) [inline plutus-ledger/src/Ledger/Tx/Internal.hs:111]: Why not just change the representation of `TxIn` instead of introducing a new type?

**Review by koslambrou — CHANGES_REQUESTED** [review]
My first pass.

TBH, this is too big a PR. I'm having a very hard time to follow all the changes. What's more, there are some refactoring changes that don't belong in `next-node`, but in `main`. The problem is that `main` is still the main developing branch and we regularly merge `main` into `next-node`. Refactoring changes like what's found here introduce lots of conflicts when merging (this definitely needs to be better documented).

I don't know if it's doable at this point, but I thinking that this PR should be splitted in multiple PRs with a clear scope of change (or stacked changes). @sjoerdvisscher Any advice?

However, very nice work with introducing these new features :) It's obvious a lot of work went into this PR.

File: plutus-chain-index-core/src/Plutus/ChainIndex/Tx.hs [hunk 12]
```diff
@@ -84,66 +86,67 @@ fromOnChainTx :: NetworkId -> OnChainTx -> ChainIndexTx
...
-    -- To determine the proper index we have to convert the plutus's `TxIn` to cardano-api `TxIn` and
-    -- sort them by using the standard `Ord` instance.
-    withHash ix (ConsumeScriptAddress _lang val red dat) =
-      let (ValidatorHash vh) = validatorHash val
-       in ( Map.singleton (ScriptHash vh) (getValidator val)
-          , Map.singleton (datumHash dat) dat
-          , Map.singleton (RedeemerPtr Spend ix) red
-          )
-    withHash _ _ = mempty
+-- TODO: the index of the txin is probably incorrect as we take it from the set.
```
[the remark this answers is not in the document]
Comment (zmrocze) [reply plutus-chain-index-core/src/Plutus/ChainIndex/Tx.hs:134]: This TODO is not my introduction, I only moved it to where appriopriate after the change. 
As I understand it is 'probably incorrect' as we should order TxIn based on the Cardano.Api ordering of inputs (maybe they're matching?).

After the change the assignment of inputs to RedeemerPtr's remains the same as before the change for spending inputs. For the rest it is only now implemented as in the code.

File: plutus-chain-index-core/src/Plutus/ChainIndex/Tx.hs [hunk 13]
```diff
@@ -84,66 +86,67 @@ fromOnChainTx :: NetworkId -> OnChainTx -> ChainIndexTx
...
+calculateRedeemerPointers :: Tx -> Redeemers
+calculateRedeemerPointers tx = spends <> rewards <> mints <> certs
+    -- we sort the inputs to make sure that the indices match with redeemer pointers
 
-txRedeemersWithHash :: ChainIndexTx -> Map RedeemerHash Redeemer
-txRedeemersWithHash ChainIndexTx{_citxRedeemers} = Map.fromList
-    $ fmap (\r -> (redeemerHash r, r))
-    $ Map.elems _citxRedeemers
+    where
+        -- These sorts are ofc bad.
```
[the remark this answers is not in the document]
Comment (zmrocze) [reply plutus-chain-index-core/src/Plutus/ChainIndex/Tx.hs:142]: okey, maybe not ofc. This applies https://github.com/input-output-hk/plutus-apps/pull/468#discussion_r945947243.

I am just unsure if in there we care if the `RedeemerPtr`'s match ones in ledger/node or only if theyre consistent across `plutus-apps`?

File: plutus-contract/src/Wallet/Emulator/Wallet.hs [hunk 14]
```diff
@@ -479,14 +479,17 @@ handleBalanceTx utxo utx = do
         logDebug $ AddingCollateralInputsFor remainingCollFees
         addCollateral utxo remainingCollFees tx''
 
+-- Auxiliary.
+newtype PubKeyTxIn = PubKeyTxIn { pubKeyTxOutRef :: TxOutRef }
```
[the remark this answers is not in the document]
Comment (zmrocze) [reply plutus-contract/src/Wallet/Emulator/Wallet.hs:483]: Yes, will change to type synonym, felt bad leaving simply as `TxOutRef`

File: plutus-ledger/src/Ledger/Tx/Internal.hs (hunk 11 above)
Comment (zmrocze) [reply plutus-ledger/src/Ledger/Tx/Internal.hs:111]: `TxIn` remains used, wanted to make the change the easiest way. Imo best in the future is to remove `Maybe` in the `txInType` field of `TxIn` (like  `TxInputWitnessed`). I haven't found this field being optional as needed.

File: plutus-ledger/src/Ledger/Tx.hs (hunk 8 above)
Comment (zmrocze) [reply plutus-ledger/src/Ledger/Tx.hs:107]: idk, moved back

File: plutus-chain-index-core/src/Plutus/ChainIndex/Tx.hs [hunk 15]
```diff
@@ -84,66 +86,67 @@ fromOnChainTx :: NetworkId -> OnChainTx -> ChainIndexTx
...
+        rewards = Map.fromList $ zipWith (\n (_, rd) -> (RedeemerPtr Reward n, rd)) [0..]  $ sort $ Map.assocs $ txRewardingRedeemers tx
+        mints   = Map.fromList $ zipWith (\n (_, rd) -> (RedeemerPtr Mint n, rd)) [0..]  $ sort $ Map.assocs $ txMintingRedeemers tx
+        certs   = Map.fromList $ zipWith (\n (_, rd) -> (RedeemerPtr Cert n, rd)) [0..]  $ sort $ Map.assocs $ txCertifyingRedeemers tx
+        -- This is written this way not to change the previous semantic. Either though previous version was commented as "probably incorrect".
```
[the remark this answers is not in the document]
Comment (zmrocze) [reply plutus-chain-index-core/src/Plutus/ChainIndex/Tx.hs:146]: Sorry, but I placed this for review only. At first I thought that the previous implementation was temporary. I wrote mine, different, and thought it didn't matter here. It did, 2 tests failed, so I recreated the previous implementation and marked it with this comment.

I removed the comments now.

File: plutus-ledger/src/Ledger/Generators.hs (hunk 7 above)
Comment (zmrocze) [reply plutus-ledger/src/Ledger/Generators.hs:212]: tldr: introduced as least friction fix

Wanted to leave implementation of `genValidTransactionSpending` that was using `TxIn`s. `txInType` in `TxIn` is optional, which made it impossible as its not optional in `TxInput`. The code can remain without big rewrite using `TxIn` without optionality, that is `TxInputWitnessed`.

File: plutus-chain-index-core/src/Plutus/ChainIndex/Tx.hs (hunk 13 above)
[the remark this answers is not in the document]
Comment (koslambrou) [reply plutus-chain-index-core/src/Plutus/ChainIndex/Tx.hs]: I meant, what does the `ofc` acronym mean?

[the remark this answers is not in the document]
Comment (zmrocze) [reply plutus-chain-index-core/src/Plutus/ChainIndex/Tx.hs]: sorry for confusion. ofc="of course" - as the ordering is different than it would be in cardano-ledger etc.

sorry for confusion, I only need to now if:
 1. assignment of RedeemerPtr's needs match the one from some formal spec
 2. assignment of RedeemerPtr's only needs to be consistent across plutus-apps code?
 
Also am I even right thinking the assignment of RedeemerPtr's for spending redeemers only satisfied 2.?

[the remark this answers is not in the document]
Comment (koslambrou) [reply plutus-chain-index-core/src/Plutus/ChainIndex/Tx.hs]: @ak3n @sjoerdvisscher

[the remark this answers is not in the document]
Comment (sjoerdvisscher) [reply plutus-chain-index-core/src/Plutus/ChainIndex/Tx.hs]: This is fine since this is only relevant for emulator Tx's.

[the remark this answers is not in the document]
Comment (zmrocze) [reply plutus-chain-index-core/src/Plutus/ChainIndex/Tx.hs]: thats great

**Comment by zmrocze** [comment]
Hey @koslambrou @ak3n, what is the plan with this pr? 

I've fixed requested bits and can update it again if you are ready. Is there something more fundamentally wrong in the changes? If so let's discuss it as well (there's some suitable issues linked at the top).

As to splitting it into smaller pr's - to me this seems like unnecessary work, I'd prefer not to do it. I'll summarize changes in squashed commit, is it alright?

**Comment by koslambrou** [comment]
Conflicts need to be fixed first, then CI needs to pass.

And yes, please summarize the changes in the squash commit.

**Review by sjoerdvisscher — APPROVED** [review]
Ok, I think I can live with merging this now to get this over with. But there needs to be a big refactor afterwards to get rid of having 3 different `TxIn` variants.

**Comment by zmrocze** [comment]
given how long lived this branch is I can't reasonably squash the commits. Doing so results in conflicts with `next-node` which were resolved already. I know it took long, sorry

**Comment by zmrocze** [comment]
Here's a description of changes if that's needed:
```
Add missing fields to Tx.

  * Add certificates and withdrawals to Tx.

  * Move script witnesses to txScripts.

  * Modify TxIn type, rename to TxInput.

  * Put redeemers together with minting scripts.

  * Translate withdrawals in toCardanoTxBody.

  * Export redeemers for signing with wallet.

```

**Review by koslambrou — APPROVED** [review]
